### PR TITLE
Add security and serialization documentation [API-635] [API-636]

### PR DESCRIPTION
**Pull request by zpqrtbnk**

### CODE REVIEW

File: doc/dev/doc/serialization.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+Let's say your custom `CustomSerializer` will serialize `CustomSerializableType`.
+
+```csharp
+public class CustomSerializer : IStreamSerializer<CustomSerializableType>
+{
+    public int TypeId => 10;
+    {
+        return 10;
+    }
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:227]:
```suggestion
    public int TypeId => 10;
```

File: doc/dev/doc/serialization.md [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+        var array = Encoding.UTF8.GetBytes(t.Value);
+        output.WriteInt(array.Length);
+        output.Write(array);
+    }
+
+    public CustomSerializableType Read(IObjectDataInput input)
+    {
+        var len = input.ReadInt();
+        var array = new byte[len];
+        input.ReadFully(array);
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:240]:
```suggestion
        input.Read(array, 0, array.Length);
```
Haven't found any `ReadFully` implementations in current codebase.

File: doc/dev/doc/serialization.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        return new CustomSerializableType { Value = Encoding.UTF8.GetString(array) };
+    }
+}
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:243]: `IStreamSerializer` inherits `IDisposable` through `ISerializer`
```suggestion
    public void Dispose()
    { }
}
```

File: doc/dev/doc/serialization.md [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+Note that the serializer `TypeId` must be unique as Hazelcast will use it to lookup the `CustomSerializer` while it deserializes the object.
+Now the last required step is to register the `CustomSerializer` to the configuration.
+
+**Programmatic Configuration:**
+
+```c#
+var hazelcastOptions = new HazelcastOptionsBuilder().Build();
+hazelcastOptions.Serialization
+    .AddSerializer(typeof(CustomSerializableType), new CustomSerializer());
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:254]: This won't compile in current version.

```suggestion
hazelcastOptions.Serialization.Serializers.Add(new SerializerOptions
{
    SerializedType = typeof(CustomSerializableType),
    Creator = () => new CustomSerializer()
});
```

File: doc/dev/doc/serialization.md [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+    Console.WriteLine($"Retrieved {result.Count} values whose age is less than 6.");
+    Console.WriteLine($"Entry is: {result.First()}");
+```
+
+## Global Serialization
+
+The global serializer is identical to custom serializers from the implementation perspective. It is registered as a fallback serializer to handle all other objects if a serializer cannot be located for them. By default, the global serializer does not handle .NET Serializable instances. However, you can configure it to be responsible for those instances.
+
+A custom serializer should be registered for a specific class type. The global serializer will handle all class types if all the steps in searching for a serializer fail as described in Serialization Interface Types.
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:310]: There's no such section as "Serialization Interface Types", does this mean all serialization interfaces described before? Why title case then?

File: doc/dev/doc/serialization.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+A sample global serializer that integrates with a third party serializer is shown below.
+
+```csharp
+public class GlobalSerializer : IStreamSerializer<object>
+{
+    public int TypeId => 20;
+
+    public void Write(IObjectDataOutput output, object obj)
+    {
+        output.write(MyFavoriteSerializer.serialize(obj))
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:326]:
```suggestion
        output.Write(MyFavoriteSerializer.Serialize(obj));
```

File: doc/dev/doc/serialization.md [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    }
+
+    public object Read(IObjectDataInput input)
+    {
+        return MyFavoriteSerializer.deserialize(input);
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:331]:
```suggestion
        return MyFavoriteSerializer.Deserialize(input);
```

File: doc/dev/doc/serialization.md [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+```
+
+**Declarative Configuration:**
+```json
+{
+    "hazelcast": {
+        "serialization": {
+            "globalSerializer": {
+                "serializedTypeName": "Hazelcast.Examples.CustomSerializableType",
+                "overrideClrSerialization": true
```
Comment (alexb5dh) [inline doc/dev/doc/serialization.md:352]: _HazelcastOptions.json_ in Tests has a bit different schema, not sure if casing matters:
```suggestion
                "typeName": "SampleNamespace.CustomSerializableType",
                "OverrideClrSerialization": true
```

File: doc/dev/doc/tlsssl.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    var sslOptions = hazelcastOptions.Networking.Ssl;
+
+    sslOptions.Enabled = true;
+    sslOptions.ValidateCertificateChain = true;
+    sslOptions.ValidateCertificateName = false;
+    sslOptions.CheckCertificateRevocation = false;
+    sslOptions.CertificateName = "CN or SAN of server certificate";
+    sslOptions.CertificatePath = "client pfx file path";
+    sslOptions.CertificatePassword = "client pfx password";
+    sslOptions.SslProtocol = SslProtocols.Tls12;
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:31]:
```suggestion
    sslOptions.Protocol = SslProtocols.Tls12;
```

File: doc/dev/doc/tlsssl.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+Of course these can also be configured via command-line options or environment variables, or via the Hazelcast configuration file. See the [configuration](configuration.md) page for details.
+
+### Enabling TLS/SSL
+
+TLS/SSL for the Hazelcast .NET client can be enabled/disabled using the `Enabled` option. When this option is set to `true`, TLS/SSL will be configured with respect to the other `SslOptions` options. Setting this option to `false` will result in discarding the other `SslOptions` options.
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:38]:
```suggestion
TLS/SSL for the Hazelcast .NET client can be enabled/disabled using the `Enabled` option. When this option is set to `true`, TLS/SSL will be configured with respect to the other `SslOptions` options. Setting this option to `false` will result in discarding other `SslOptions` properties.
```

File: doc/dev/doc/tlsssl.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+Default value is `false` (disabled). 
+
+### Certificate Chain validation
+
+Remote SSL certificate chain validation can be enabled/disabled using the `SslOptions.ValidateCertificateChain` option. It is enabled by default. If you need to bypass certificate validation for some reason, you can disable it as follows by setting the value to `false`.
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:44]:
```suggestion
Remote SSL certificate chain validation can be enabled/disabled using the `SslOptions.ValidateCertificateChain` option. It is enabled by default. If you need to bypass certificate validation for some reason, you can disable it by setting the value to `false`. 
```

File: doc/dev/doc/tlsssl.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+Validation is done by .NET and delegated to OS, and you need to make sure your server certificate is trusted by your OS.
+Please refer to [this blog](https://blogs.msdn.microsoft.com/webdev/2017/11/29/configuring-https-in-asp-net-core-across-different-platforms/) for information on how to configure your OS to trust your server certificates.
+
+### Certificate Name Validation
+
+Server certificate CN or SAN field can be validated against a value you set into configuration. This option is disabled by default. You can enable it by setting `SslOptions.ValidateCertificateName` to `true` and providing a name with `SslOptions.CertificateName`.
+
+### TLS/SSL Protocol
+
+You can configure the TLS/SSL protocol using the `SslOptions.Protocol` option. Valid options are string values of `System.Security.Authentication.SslProtocols` Enum. Depending on your .Net Framework/Net core version, below values are valid:
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:55]:
```suggestion
You can configure the TLS/SSL protocol using the `SslOptions.Protocol` option. Valid options are values of `System.Security.Authentication.SslProtocols` enum. Depending on your .NET Framework/.NET Core version, below values are valid:
```

File: doc/dev/doc/tlsssl.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+* **None**    : Allows the operating system to choose the best protocol to use. 
+* **Ssl2**    : SSL 2.0 Protocol. *RFC 6176 prohibits the usage of SSL 2.0.* 
+* **Ssl3**    : SSL 3.0 Protocol. *RFC 7568 prohibits the usage of SSL 3.0.*
+* **Tls**     : TLS 1.0 Protocol described in RFC 2246. *deprecated.*
+* **Tls11**   : TLS 1.1 Protocol described in RFC 4346. *deprecated.*
+* **Tls12**   : TLS 1.2 Protocol described in RFC 5246. *deprecated.*
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:62]: Deprecated for Hazelcast specifically? TLS 1.2 seems to still be quite popular version on the internet.

Also, what about `SslProtocols.Tls13`? This list makes it look like one shouldn't use anything apart from `None` option. 😕

File: doc/dev/doc/toc.md [hunk 14]
```diff
@@ -16,9 +19,10 @@
 ## [AtomicLong](atomiclong.md)
 ## [AtomicRef](atomicref.md)
 # [CP SubSystem](cpsubsystem.md)
-# [Locking](locking.md)
-# [Events](events.md)
-# [Kerberos](kerberos.md)
+# [Concepts](concepts.md)
```
Comment (alexb5dh) [inline doc/dev/doc/toc.md:22]: What is the idea for grouping these three under "Concepts" section?

File: doc/dev/doc/serialization.md [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+|---------|-----------|
+| bool    | Boolean   |
+| byte    | Byte      |
+| char    | Character |
+| short   | Short     |
+| int     | Integer   |
+| long    | Long      |
+| float   | Float     |
+| double  | Double    |
+| string  | String    |
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:21]: ok I'm adding `Guid` - I'd rather not document `KeyValuePair<,>` for now as I'm not sure what it's for, as you mentioned.

File: doc/dev/doc/serialization.md [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+
+## Serialization Priority
+
+When Hazelcast .NET client serializes an object into `IData`:
+
+1. It first checks whether the object is null.
+2. If the above check fails, then Hazelcast checks if it is an instance of `Hazelcast.Serialization.IIdentifiedDataSerializable`.
+3. If the above check fails, then Hazelcast checks if it is an instance of `Hazelcast.Serialization.IPortable`.
+4. If the above check fails, then Hazelcast checks if it is an instance of one of the default types (see above default types).
+5. If the above check fails, then Hazelcast looks for a user-specified Custom Serializer, i.e., an implementation of `IByteArraySerializer<T>` or `IStreamSerializer<T>`. Custom serializer is searched using the input object’s class and its parent class up to `Object`. If parent class search fails, all interfaces implemented by the class are also checked.
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:41]: not sure I follow you? the code (see `SerializationService.SerializerFor(...)`) does test for `IIdentifiedDataSerializable`, *then* for `IPortable`, etc - so the order documented here is guaranteed?

File: doc/dev/doc/serialization.md [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+6. If the above check fails, then Hazelcast checks if it is Serializable ( `Type.IsSerializable` ) and a Global Serializer is not registered with CLR serialization Override feature.
+7. If the above check fails, Hazelcast will use the registered Global Serializer if one exists.
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:43]: adding a note

File: doc/dev/doc/serialization.md [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    }
+}
+```
+
+The last step is to register the `IDataSerializableFactory` to the `SerializationOptions`.
+
+**Programmatic Configuration:**
+```csharp
+var hazelcastOptions = new HazelcastOptionsBuilder().Build();
+var factory = new Hazelcast.Examples.SampleDataSerializableFactory();
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:95]: yes, fixing everywhere

File: doc/dev/doc/serialization.md [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+
+**Declarative Configuration:**
+```json
+{
+    "hazelcast": {
+        "serialization": {
+            "dataSerializableFactories": [
+                {
+                    "id": 1000,
+                    "typeName": "Hazelcast.Examples.SampleDataSerializableFactory"
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:108]: yes, adding a note

File: doc/dev/doc/tlsssl.md (hunk 10 above)
Comment (zpqrtbnk) [reply doc/dev/doc/tlsssl.md:38]: yes

File: doc/dev/doc/tlsssl.md (hunk 11 above)
Comment (zpqrtbnk) [reply doc/dev/doc/tlsssl.md:44]: yes

File: doc/dev/doc/tlsssl.md (hunk 12 above)
Comment (zpqrtbnk) [reply doc/dev/doc/tlsssl.md:55]: yes

File: doc/dev/doc/tlsssl.md (hunk 13 above)
Comment (zpqrtbnk) [reply doc/dev/doc/tlsssl.md:62]: my mistake, `Tls12` is recommended, not deprecated, fixing

File: doc/dev/doc/toc.md (hunk 14 above)
Comment (zpqrtbnk) [reply doc/dev/doc/toc.md:22]: i am trying to structure the doc somehow... these felt like "leftovers" but yea I don't like "concepts" much and then why is serialization not a concept? ok, killing "concepts" for now ;)

**Comment by zpqrtbnk** [comment]
have pushed a commit with fixes

File: doc/dev/doc/serialization.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (alexb5dh) [reply doc/dev/doc/serialization.md:42]: I'm talking about this sentence:
> If parent class search fails, all interfaces implemented by the class are also checked.

First matching interface found will guide the serialization logic, but order of interfaces doesn't seem to be guaranteed in any way.

File: doc/dev/doc/tlsssl.md [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+Validation is done by .NET and delegated to OS, and you need to make sure your server certificate is trusted by your OS.
+Please refer to [this blog](https://blogs.msdn.microsoft.com/webdev/2017/11/29/configuring-https-in-asp-net-core-across-different-platforms/) for information on how to configure your OS to trust your server certificates.
+
+### Certificate Name Validation
+
+Server certificate CN or SAN field can be validated against a value you set into configuration. This option is disabled by default. You can enable it by setting `SslOptions.ValidateCertificateName` to `true` and providing a name with `SslOptions.CertificateName`.
+
+### TLS/SSL Protocol
+
+You can configure the TLS/SSL protocol using the `SslOptions.Protocol` option. Valid options are values of the `System.Security.Authentication.SslProtocols` enum. Depending on your .Net version, below values are valid:
```
Comment (alexb5dh) [inline doc/dev/doc/tlsssl.md:55]: It's usually capitalized and we have it like this in other places.
```suggestion
You can configure the TLS/SSL protocol using the `SslOptions.Protocol` option. Valid options are values of the `System.Security.Authentication.SslProtocols` enum. Depending on your .NET version, below values are valid:
```

File: doc/dev/doc/serialization.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/serialization.md:42]: ah - right - so if there is no `IByteArraySerializer<T>` nor `IStreamSerializer<T>` where `T` is the object's class or any of it's ancestor classes... we're looking for eg `IStreamSerializer<T>` where `T` is any of the object's interfaces -- and the order of these interfaces is not specified - right? updating the text.

**Review by alexb5dh — APPROVED** [review]

**Pull request merged**